### ISSUE TITLE
wrong cookie parse

### ISSUE THREAD

**Issue by gitlsl** [dotnet/runtime#74794]
### Description

Cookie parsing error

### Reproduction Steps

````c#
var handler = new HttpClientHandler();
var client = new HttpClient(handler);

var t = await client.GetStringAsync("http://www.baidu.com"); //first
t = await client.GetStringAsync("http://www.baidu.com"); //second
````

### Expected behavior

second query with    cookie   which return from first time response,

I want to get same behavior like python

### Actual behavior

first
````
request:
GET http://www.baidu.com/ HTTP/1.1
Host: www.baidu.com

response:
HTTP/1.1 200 OK
Content-Length: 9508
Accept-Ranges: bytes
Cache-Control: no-cache
Connection: keep-alive
Content-Type: text/html
Date: Tue, 30 Aug 2022 06:08:26 GMT
Keep-Alive: timeout=4
P3p: CP=" OTI DSP COR IVA OUR IND COM "
P3p: CP=" OTI DSP COR IVA OUR IND COM "
Pragma: no-cache
Proxy-Connection: keep-alive
Server: BWS/1.1
Set-Cookie: BAIDUID=81534E54E34A553C75DC35D27D558FA5:FG=1; expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/; domain=.baidu.com
Set-Cookie: BIDUPSID=81534E54E34A553C75DC35D27D558FA5; expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/; domain=.baidu.com
Set-Cookie: PSTM=1661839706; expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/; domain=.baidu.com
Set-Cookie: BAIDUID=81534E54E34A553C989408629806AAD0:FG=1; max-age=31536000; expires=Wed, 30-Aug-23 06:08:26 GMT; domain=.baidu.com; path=/; version=1; comment=bd
Traceid: 1661839706062388225011011922799548172837
````

second:
````
request:
GET http://www.baidu.com/ HTTP/1.1
Host: www.baidu.com
Cookie: $Version=1; BAIDUID=81534E54E34A553C989408629806AAD0:FG=1; $Path=/; $Domain=.baidu.com


````

I also use python test
same code as above
````
first 
GET http://www.baidu.com/ HTTP/1.1
Host: www.baidu.com
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
User-Agent: python-httpx/0.23.0



second:
GET http://www.baidu.com/ HTTP/1.1
Host: www.baidu.com
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
User-Agent: python-httpx/0.23.0
Cookie: BAIDUID=2B6E3385F13B92C7143807D6C2AF11A2:FG=1; BIDUPSID=2B6E3385F13B92C79C919ACE5644495B; H_PS_PSSID=36552_36626_36885_34812_36570_36786_37136_26350_37205; PSTM=1661845393; BDSVRTM=0; BD_HOME=1


this is right

````

### Configuration

.net core 6.0
win11 x64
console app

**Comment by skyoxZ** [comment]
Seems regression from .NET 4.8

Test the code below for .NET 4.8 and .NET 6.0:
```csharp
using System.Net.Http;

HttpClient client = new HttpClient();
client.GetStringAsync("http://www.baidu.com").Wait();
client.GetStringAsync("http://www.baidu.com").Wait();
```
and I use Fiddler to capture the cookie of the second request in each test.

Result:

.NET 4.8:
```
Cookie: $Version=1; BAIDUID=E292CDA7293B7796CD8C86B71709E50C:FG=1; $Path=/; $Domain=.baidu.com; BIDUPSID=E292CDA7293B77965C81F98FB0BAE5CE; PSTM=1661848401
```

.NET 6.0
```
Cookie: $Version=1; BAIDUID=7D9A7F91C42249D6990D330FD818EB6F:FG=1; $Path=/; $Domain=.baidu.com
```

I have no idea why some keys prefixed with `$` (such as `$Version`) are in the cookie but at least the result of .NET 4.8 contains all necessary key/values.

**Comment by ManickaP** [comment]
Hi @skyoxZ, could you test this on the recent .NET 7.0 preview version? There were some changes in this space so it might already be fixed. If not, we should investigate and fix this.

**Comment by skyoxZ** [comment]
@ManickaP .NET 7 preview7 results same with .NET 6.

**Comment by antonfirsov** [comment]
A smaller repro:

```C#
string[] rawCookies = {
    "BAIDUID=9C50A30DD72E13C3C693127F24BC25E0:FG=1; expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/; domain=.baidu.com",
    "BIDUPSID=9C50A30DD72E13C3C693127F24BC25E0; expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/; domain=.baidu.com",
    "PSTM=1662408874; expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/; domain=.baidu.com",
    "BAIDUID=9C50A30DD72E13C3619BE7A3B2840D68:FG=1; max-age=31536000; expires=Tue, 05-Sep-23 20:14:34 GMT; domain=.baidu.com; path=/; version=1; comment=bd"
};

Uri uri = new Uri("http://www.baidu.com");

CookieContainer container = new CookieContainer();
foreach (string c in rawCookies)
{
    container.SetCookies(uri, c);
}

Console.WriteLine("Count: " + container.Count);
```

Starting from .NET 5, this code prints 1 instead of 3. I can't immediately tell what is happening, but Chrome seems to keep the cookies we drop, so it's worth an investigation.

**Comment by gitlsl** [comment]
@antonfirsov  I  change `31-Dec-37` to `31-Dec-27`  , your code return  `count:3 `   ,   also  I have no idea why some keys prefixed with ` $ `

**Comment by danmoseley** [comment]
https://github.com/dotnet/runtime/blob/e55c908229e36f99a52745d4ee85316a0e8bb6a2/src/libraries/Common/src/System/Net/CookieParser.cs#L649

"31-Dec-37" is parsed as "12/31/1937" as CultureInfo.InvariantCulture.Calendar.TwoDigitYearMax = 29

**Comment by danmoseley** [comment]
BTW does this

https://github.com/dotnet/runtime/blob/e55c908229e36f99a52745d4ee85316a0e8bb6a2/src/libraries/Common/src/System/Net/CookieParser.cs#L663

Imply that the first of Expires and Max-Age wins? As the RFC (4.1.2.2) says if both are supplied then Max-Age should win.

https://www.rfc-editor.org/rfc/rfc6265

**Comment by tarekgh** [comment]
@danmoseley when running on ICU we don't read the system settings https://github.com/dotnet/runtime/blob/2a747212d2bf60b633a567b8992b06ceab3bb85f/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Icu.cs#L88. 

It looks like this was the original behavior when started using ICU as we were supporting non-Windows platforms only with ICU. Nothing prevent us now to try to read these settings now when running on Windows as we do here https://github.com/dotnet/runtime/blob/2a747212d2bf60b633a567b8992b06ceab3bb85f/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Nls.cs#L19.

In your case you are getting 2029 as this is the default value when using ICU https://github.com/dotnet/runtime/blob/2a747212d2bf60b633a567b8992b06ceab3bb85f/src/libraries/System.Private.CoreLib/src/System/Globalization/GregorianCalendar.cs#L409. 

In general, we can update these numbers, but we need to consider it as a breaking change.

**Comment by danmoseley** [comment]
OK, I'll open an issue to consider, as this is quite separate from what if anything is decided for cookies here.

**Comment by danmoseley** [comment]
Opened https://github.com/dotnet/runtime/issues/75148. Incidentally I noticed [RFC3339](https://www.rfc-editor.org/rfc/rfc3339#page-4) says 

>  Internet Protocols MUST generate four digit years in dates.

> The use of 2-digit years is deprecated.  If a 2-digit year is
         received, it should be accepted ONLY if an incorrect
         interpretation will not cause a protocol or processing failure
         (e.g. if used only for logging or tracing purposes).

**Comment by karelz** [comment]
Triage: Not a regression, moving to 8.0. Not clear if anything has to be done -- @antonfirsov can hopefully shed some light when he is back.

**Comment by antonfirsov** [comment]
This should be fixed with #76848.
